### PR TITLE
cleanup test images

### PR DESCRIPTION
**Pull request by jessfraz**
- remove jess/unshare
- add more seccomp tests

so I removed an image and added one :innocent: 

ping @tianon @justincormack

### CODE REVIEW

**Comment by tianon** [comment]
Arg, you're killing me, Jess :heart:

**Comment by jfrazelle** [comment]
the other one is better tho :)

**Comment by jfrazelle** [comment]
I could write c tests for those instead i debated on it, would be easy

**Comment by jfrazelle** [comment]
actually ill do that, it will be easier

**Comment by justincormack** [comment]
It is probably easiest to write tests that would fail with EFAULT or EINVAL, ie make most of the parameters null, as EPERM is easy to distinguish, then you dont need to care what the parameters are. Do not however set your clock to null like I did accidentally earlier today while testing, it makes SSL very unhappy.

**Comment by tianon** [comment]
WRT multi-arch, it'd be easiest if you could write something that is compiled/built as part of the tests (similar to our `ensure-http-server`) based on one of the official images that are already known to be easy to port (like `debian:jessie` or similar) so we don't have to build another one-off image multi-arch as part of the official stuff.

**Comment by tianon** [comment]
(I've got `alpine` working on `armhf`, but Alpine upstream doesn't support `ppc64le`, `s390x`, `arm64`, or anything outside of `amd64`, `i386`, and `armhf` for that matter, so something that builds off `debian:jessie` or newer would be better.)

**Comment by justincormack** [comment]
I could base `justincormack/ljsyscall` off `debian:jessie` rather than `alpine` it supports all those except no idea about `s390x` (its some sort of powerpc?). I did say I would port Musl libc to `ppc64le` but havent yet. There is an `arm64` port, so hopefully Alpine soon.

**Comment by jfrazelle** [comment]
nah im almost done :)

**Comment by jfrazelle** [comment]
updated :)

**Comment by justincormack** [comment]
the `acct` test will fail with ENOFILE with `/tmp/t` as an argument, not succeed I think.

**Comment by jfrazelle** [comment]
ya i took that into account, pun intended

**Comment by tianon** [comment]
@justincormack https://en.wikipedia.org/wiki/IBM_System_z :disappointed: (crazy IBM arch -- reasonably performant if you can get your hands on one, but not sure who has them besides IBM)

See also https://wiki.debian.org/ArchiveQualification/s390x and https://fedoraproject.org/wiki/Architectures/s390x.

Getting musl and thus alpine on ppc64le and arm64 would be really epic though. :smile:

File: hack/make/.ensure-syscall-test [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+# and compile it for target daemon
+
+dir="$DEST/syscall-test"
+mkdir -p "$dir"
+(
+	if [ "$(go env GOOS)" = "linux" ]; then
+		cd "$dir"
+		gcc -g -Wall -static ../../../../contrib/syscall-test/userns.c -o ./userns-test
+		gcc -g -Wall -static ../../../../contrib/syscall-test/ns.c -o ./ns-test
+		gcc -g -Wall -static ../../../../contrib/syscall-test/acct.c -o ./acct-test
```
Comment (tianon) [inline hack/make/.ensure-syscall-test]: Do you think it's worth installing `gcc` in the container so we can compile and run these tests properly on a host whose arch doesn't match the one we're using? :smile:

Comment (jfrazelle) [inline hack/make/.ensure-syscall-test]: sure thing

**Comment by justincormack** [comment]
Most architectures need aligned stacks, so the child stack may fail. I would use syscall(SYS_CLONE, ...) and leave the stack null (see the note at the bottom of the man page), or allocate the mmap(... MAP_STACK, ...).

File: hack/make/.ensure-syscall-test (hunk 1 above)
Comment (tianon) [inline hack/make/.ensure-syscall-test]: You're a hero :heart: :heart:

**Comment by jfrazelle** [comment]
I used `malloc` hopefully thats ok?

File: hack/make/.ensure-syscall-test [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+set -e
+
+# Build a C binary for cloning a userns for seccomp tests
+# and compile it for target daemon
+(
+	if [ "$(go env GOOS)" = "linux" ]; then
```
Comment (tianon) [inline hack/make/.ensure-syscall-test]: Shouldn't this line be using the env variable for the Docker Engine `GOOS` (`DOCKER_ENGINE_GOOS`)? (sorry, didn't notice this one before)

Comment (jfrazelle) [inline hack/make/.ensure-syscall-test]: nice good catch

**Comment by justincormack** [comment]
No alignment guarantee in malloc, you need `posix_memalign` or `stack = mmap(NULL, STACKSIZE, PROT_READ | PROT_WRITE,  MAP_SHARED | MAP_ANON | MAP_STACK, -1, 0);`

File: hack/make/.ensure-syscall-test [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+		cd contrib/syscall-test
+		docker build -qt syscall-test . > /dev/null
```
Comment (tianon) [inline hack/make/.ensure-syscall-test]: Oh, and you could just do `docker build -qt syscall-test contrib/syscall-test` and skip the `cd` and thus subshell entirely. :innocent:

**Comment by justincormack** [comment]
LGTM

**Comment by tianon** [comment]
LGTM :heart: :heart:

**Comment by aaronlehmann** [comment]
Would it be possible to build the test image as part of the top-level Dockerfile instead of doing it at test-integration-cli time? With this PR, every attempt to run a test to creates the image from scratch, which involves downloading and installing a C compiler. It slows down the test development cycle, and means tests can never be run offline.

**Comment by tianon** [comment]
Hmm, not easily, but we could vendor the "gcc" image and use that as the
base instead which would speed the "docker build" bit up a fair amount.

**Comment by aaronlehmann** [comment]
@tianon: I don't understand why the image needs to be built every time the tests are run. I was thinking the image could be built as a step in the top-level Dockerfile, so it's available for all future test invocations inside the container where tests run. Are there issues with doing nested `docker build`s?

I'd really appreciate anything you can do to improve this. It's been pretty painful today when I was working on some tests. The ensure-syscall-test bundle seems to add about 25s to every test invocation, and I have a very fast internet connection. For me it's merely an annoyance - for someone with a slow internet connection or a restrictive data cap, it could make Docker development impossible.

**Comment by tianon** [comment]
Yeah -- we don't have a Docker daemon available inside the containers
created during "RUN" lines in the Dockerfile, so we can't easily "docker
build" there, but switching to the slightly larger "gcc" image should work
well enough.

**Comment by aaronlehmann** [comment]
Alternatively, if the docker-dev image created by the top-level Dockerfile is available (or can be imported), the syscall-test image could be based on that (it appears to include gcc). Not sure if this would be overly complicated.

**Comment by tianon** [comment]
We run from inside that image (and we don't actually require it being built
at all), so unfortunately that's not viable. :disappointed:

**Comment by aaronlehmann** [comment]
How about this:

- The test-integration-cli bundle compiles the C programs and links them statically (this runs inside the docker-dev container or an environment with similar dependencies, so gcc should be available)
- The syscall-test image is built `FROM scratch`
- When the syscall-test image is run, whatever runs it uses `docker cp` and `docker exec` to run the test programs inside the container.

**Comment by jfrazelle** [comment]
but then different arches have to make these images

**Comment by aaronlehmann** [comment]
Not sure I get it. Are the tests ever running on a different architecture than the build system (except Windows, where I don't think this is applicable)?

**Comment by jfrazelle** [comment]
arm, powerpc, z

**Comment by jfrazelle** [comment]
oh i didnt see the message about compiling and copying sorry

**Comment by tianon** [comment]
In some cases, yes; for example, running the tests on Windows against a
remote Linux daemon (used for testing the client only functionality).

**Comment by BrianBland** [comment]
:+1: for @aaronlehmann's suggestion of compiling the C programs in the docker-dev container. Am I wrong in assuming that this should be able to cross-compile the binaries for use with any architecture combination?

**Comment by tianon** [comment]
It's possible, but since these are C binaries would require complicated
cross-compiler setup, which is easy to encode in our Dockerfile, but not
easily replicated to places like the Windows CI workers, for example.

**Comment by tianon** [comment]
I'm working on a PR to make this fast again, just give me a little bit
to test it properly. :smile:

**Comment by tianon** [comment]
See https://github.com/docker/docker/pull/19345 :+1:

**Pull request merged**